### PR TITLE
Allow HTTP pushes directly to ingesters.

### PR DESCRIPTION
**Pull request by tomwilkie**
This is mostly for benchmark comparisons.

Signed-off-by: Tom Wilkie <tom.wilkie@gmail.com>

### CODE REVIEW

**Comment by bboreham** [comment]
We should drop that "billing" stuff.  Unfortunately I probably need to put something else in its place or I'll spook the natives.

**Comment by tomwilkie** [comment]
Agreed re: billing.  We use the per-user metrics for billing now.

**Comment by tomwilkie** [comment]
@bboreham I've remove the billing code, WDYT?

**Comment by bboreham** [comment]
I’m still at “ need to put something else in its place”. 
Also you should mention this change in the title and in changelog.

**Comment by tomwilkie** [comment]
> I’m still at “ need to put something else in its place”.

Do you want us to wait?  When do you think you'll have a chance to do this?

File: pkg/cortex/modules.go [hunk 1]
```diff
@@ -341,6 +342,7 @@ func (t *Cortex) initIngester(cfg *Config) (err error) {
 	t.server.HTTP.Path("/ready").Handler(http.HandlerFunc(t.ingester.ReadinessHandler))
 	t.server.HTTP.Path("/flush").Handler(http.HandlerFunc(t.ingester.FlushHandler))
 	t.server.HTTP.Path("/shutdown").Handler(http.HandlerFunc(t.ingester.ShutdownHandler))
+	t.server.HTTP.Handle("/push", t.httpAuthMiddleware.Wrap(push.Handler(cfg.Distributor, t.ingester.Push)))
```
Comment (pracucci) [inline pkg/cortex/modules.go]: Two things:
1. I'm a bit concerned about the inconsistency in the path between distributors (`/api/prom/push`) and ingesters (`/push`). Have you thought about it?
2. Could you document it to `docs/apis.md`, please?

**Comment by bboreham** [comment]
> > I’m still at “ need to put something else in its place”.
> Do you want us to wait? When do you think you'll have a chance to do this?

I have written the program.  I don't think it's essential that you wait.

File: pkg/cortex/modules.go (hunk 1 above)
Comment (tomwilkie) [reply pkg/cortex/modules.go]: > I'm a bit concerned about the inconsistency in the path between distributors (/api/prom/push) and ingesters (/push). Have you thought about it?

Yeah, we can have them at the same path as the paths will collide in the single binary case.

I put it at `/push` to indicate its a private API; the /api/prom is for public endpoints.

**Pull request merged**